File: lua/glacier/widget/prompt.lua
local Widget = require("snowcap.widget")
local StdSig = require("snowcap.widget.signal")

local color = require("glacier.misc.color")

---Glacier's Prompts
---
---@class glacier.widget.prompt
---@field mt metatable The module's metatable
---
---@overload fun(...: glacier.widget.prompt.Config): glacier.widget.Prompt
local prompt = { mt = {} }

----------------------
-- Type Definitions --
----------------------

---Signals emitted by Prompts.
---
---@enum glacier.widget.prompt.signal
local _signal = {
    DONE = "glacier::widget::prompt::done",
}

---Function called when the Prompt is submitted.
---@alias glacier.widget.prompt.ExeCallback fun(input: string)

---Prompt's events
---@class glacier.widget.prompt.Event
---@field activate? {}
---@field deactivate? {}
---@field focus? {}
---@field input? string
---@field submit? {}

---Prompt's messages.
---@class glacier.widget.prompt.Message
---@field id integer
---@field event glacier.widget.prompt.Event
local Message = {}

---Prompt's input area appearance
---@class glacier.widget.prompt.InputStyle
---Color used to render text.
---@field fg_color? snowcap.widget.Color
---Background color used for the prompt
---@field bg_color? snowcap.widget.Color
---Border for the Prompt's text_input
---@field border? snowcap.widget.Border
---Color used to render the icon
---@field icon_color? snowcap.widget.Color
---Color used for the placeholder text
---@field placeholder_color? snowcap.widget.Color
---Color used for selected text
---@field selection_color? snowcap.widget.Color

---Prompt widget style.
---@class glacier.widget.prompt.Style: glacier.widget.prompt.InputStyle
---Font used to render the prompt
---@field font? snowcap.widget.Font
---Icon to display on the left or right side of the Prompt
---@field icon? snowcap.widget.text_input.Icon
---Prompt's internal padding
---@field padding? snowcap.widget.Padding
---Style overrides for active Prompt.
---@field active? glacier.widget.prompt.InputStyle
---Style overrides for focused Prompt.
---@field focused? glacier.widget.prompt.InputStyle
local Style = {}

---Prompt's configuration.
---@class glacier.widget.prompt.Config
---@field placeholder? string
---@field style? glacier.widget.prompt.Style
---@field on_exe? glacier.widget.prompt.ExeCallback

---Prompt widget.
---
---@class glacier.widget.Prompt: snowcap.widget.Program
---@field placeholder string
---@field on_exe glacier.widget.prompt.ExeCallback
---@field private _style glacier.widget.prompt.Style
---@field private _content string
---@field private _id string
---@field private _active boolean
local Prompt = setmetatable({}, { __index = require("snowcap.widget.base").Base })

----------------------
-- Module functions --
----------------------

function prompt.default_view() end

---Default `Prompt` appearance.
---
---@param style? glacier.widget.prompt.Style
---@return glacier.widget.prompt.Style
function prompt.default_style(style)
    ---@type glacier.widget.prompt.Style
    return require("glacier.utils").merge_table({
        font = {
            family = Widget.font.family.Monospace,
            weight = Widget.font.weight.SEMIBOLD,
        },
        icon = {
            code_point = utf8.codepoint(""),
            spacing = 4.0,
        },
        bg_color = color.from_hex("#000000", 0.0),
        border = {
            width = 0.0,
        },
        padding = { bottom = 0., left = 0., right = 0., top = 0. },
    }, style)
end

---Default prompt execution callback.
---
---This function split the string on whitespace, then call Process::spawn on the resulting
---array.
---
---Spawn a new process.
---@param input string
function prompt.spawn(input)
    if not input or string.len(input) == 0 then
        return
    end

    local cmd = {}

    for word in string.gmatch(input, "[^%s]+") do
        table.insert(cmd, word)
    end

    if #cmd == 0 then
        return
    end

    require("pinnacle.process").spawn(cmd)
end

-----------------------
-- Message functions --
-----------------------

---@package
---Build a focus message
---
---@param id integer
---@return glacier.widget.prompt.Message
function Message.focus(id)
    ---@type glacier.widget.prompt.Message
    return {
        id = id,
        event = {
            focus = {},
        },
    }
end

---@package
---Build an activate message
---
---@param id integer
---@return glacier.widget.prompt.Message
function Message.activate(id)
    ---@type glacier.widget.prompt.Message
    return {
        id = id,
        event = {
            activate = {},
        },
    }
end

---@package
---Build a deactivate message
---
---@param id integer
---@return glacier.widget.prompt.Message
function Message.deactivate(id)
    ---@type glacier.widget.prompt.Message
    return {
        id = id,
        event = {
            deactivate = {},
        },
    }
end

---@package
---Build an input message.
---
---@param id integer
---@param input string
---@return glacier.widget.prompt.Message
function Message.input(id, input)
    ---@type glacier.widget.prompt.Message
    return {
        id = id,
        event = {
            input = input,
        },
    }
end

---@package
---Build a submit message.
---
---@param id integer
---@return glacier.widget.prompt.Message
function Message.submit(id)
    ---@type glacier.widget.prompt.Message
    return {
        id = id,
        event = {
            submit = {},
        },
    }
end

-------------------
-- Style methods --
-------------------

---@package
---
---@param input_style glacier.widget.prompt.InputStyle
---@return snowcap.widget.text_input.Style
local function _to_textinput_style(input_style)
    ---@type snowcap.widget.text_input.Style
    return {
        background = input_style.bg_color and Widget.background.Color(input_style.bg_color) or nil,
        border = input_style.border,
        icon = input_style.icon_color,
        placeholder = input_style.icon_color,
        selection = input_style.selection_color,
        value = input_style.fg_color,
    }
end

---@private
function Style:get_active()
    local deep_copy = require("snowcap.util").deep_copy
    local merge_table = require("glacier.utils").merge_table

    ---@type glacier.widget.prompt.InputStyle
    local style = deep_copy({
        fg_color = self.fg_color,
        bg_color = self.bg_color,
        border = self.border,
        icon_color = self.icon_color,
        placeholder_color = self.placeholder_color,
        selection_color = self.selection_color,
    })

    if self.active then
        style = merge_table(style, deep_copy(self.active))
    end

    return _to_textinput_style(style)
end

---@private
function Style:get_focused()
    local deep_copy = require("snowcap.util").deep_copy
    local merge_table = require("glacier.utils").merge_table

    ---@type glacier.widget.prompt.InputStyle
    local style = deep_copy({
        fg_color = self.fg_color,
        bg_color = self.bg_color,
        border = self.border,
        icon_color = self.icon_color,
        placeholder_color = self.placeholder_color,
        selection_color = self.selection_color,
    })

    if self.focused then
        style = merge_table(style, deep_copy(self.focused))
    end

    return _to_textinput_style(style or {})
end

---@package
---
---@return snowcap.widget.text_input.Styles
function Style:to_textinput()
    ---@type snowcap.widget.text_input.Styles
    return {
        active = self:get_active(),
        focused = self:get_focused(),
        -- If the UI take too much time to refresh, iced sets the input as disabled, which doesn't look good.
        disabled = self:get_active(),
    }
end

--------------------
-- Style lifetime --
--------------------

---@package
---
---Initialize a new `Style`
---@param style glacier.widget.prompt.Style
---@return glacier.widget.prompt.Style
function Style.new(style)
    style = style or {}
    return setmetatable(style, { __index = Style })
end

---------------------------
-- Prompt public methods --
---------------------------

function Prompt:focus()
    self:emit(StdSig.send_message, Message.focus(self:id()))
end

function Prompt:activate()
    self:emit(StdSig.send_message, Message.activate(self:id()))
end

function Prompt:deactivate()
    self:emit(StdSig.send_message, Message.deactivate(self:id()))
end

----------------------------
-- Prompt private methods --
----------------------------

function Prompt:on_focus()
    if not self._active then
        return
    end

    self:emit(StdSig.operation, Widget.operation.focusable.Focus(self._id))
end

function Prompt:on_activate()
    if self._active then
        return
    end

    self._active = true
    self:on_focus()
end

function Prompt:on_deactivate()
    if not self._active then
        return
    end

    self._content = ""
    self._active = false
    self:emit(_signal.DONE, tostring(self))
end

---@private
---Extract an event from the message.
---
---@param msg any|glacier.widget.prompt.Message?
---@return glacier.widget.prompt.Event?
function Prompt:get_message_event(msg)
    if msg and msg.id == self:id() then
        return msg.event
    end

    return nil
end

function Prompt:__tostring()
    return ("<Prompt#%d>"):format(self:id())
end

---------------------------------
-- impl snowcap.widget.Program --
---------------------------------

---Creates a widget definition for display by Snowcap.
---
---A widget may return nil to notify its parent program that it has
---nothing to display. It's up to the parent to decide whether to display a
---placeholder or to remove the widget from the tree.
---@return snowcap.widget.WidgetDef?
function Prompt:view()
    if not self._active then
        return nil
    end

    local ti = Widget.text_input({
        id = self._id,
        placeholder = self.placeholder or "",
        value = self._content or "",
        padding = self._style.padding,
        font = self._style.font,
        icon = self._style.icon,
        style = self._style:to_textinput(),
        on_input = function(input)
            return Message.input(self:id(), input)
        end,
        on_submit = Message.submit(self:id()),
    })

    return Widget.container({
        child = ti,
        height = Widget.length.Fill,
        width = Widget.length.Fill,
        valign = Widget.alignment.CENTER,
    })
end

---Updates this widget program with the received message.
---@param msg any|glacier.widget.prompt.Message
function Prompt:update(msg)
    local event = self:get_message_event(msg)

    if not event then
        return
    end

    if event.activate then
        self:on_activate()
    elseif event.deactivate then
        self:on_deactivate()
    elseif event.focus then
        self:on_focus()
    elseif event.input then
        self._content = event.input
    elseif event.submit then
        if self.on_exe then
            self.on_exe(self._content)
        else
            prompt.spawn(self._content)
        end

        self:on_deactivate()
    end
end

---Called when a surface has been created with this program.
---
---A surface handle is provided to allow the program to manupulate
---the surface. This handle should be passed to any child programs
---to allow them to use it as well.
---
---@param event snowcap.widget.SurfaceEvent
function Prompt:event(event)
    if event.focus_lost then
        self:on_deactivate()
    end
end

-----------
-- Other --
-----------

---Create a new Prompt widget.
---
---@param config glacier.widget.prompt.Config
---@return glacier.widget.Prompt
function Prompt:new(config)
    config = config or {}
    config.style = require("glacier.utils").merge_table(prompt.default_style(), config.style)

    local base = require("snowcap.widget.base").Base.new()
    local ret = setmetatable(base, {
        __index = Prompt,
        __tostring = Prompt.__tostring,
    }) --[[@as glacier.widget.Prompt]]

    ret._content = ""
    ret.placeholder = config.placeholder
    ret.on_exe = config.on_exe
    ret._style = Style.new(config.style)
    ret._id = tostring(ret)
    ret._active = false

    return ret
end

---Create a new Prompt widget.
---
---@param ... glacier.widget.prompt.Config
---@return glacier.widget.Prompt
function prompt.mt:__call(...)
    return Prompt:new(...)
end

prompt.Prompt = Prompt
prompt.signal = _signal

---@diagnostic disable-next-line:param-type-mismatch
return setmetatable(prompt, prompt.mt) --[[@as glacier.widget.prompt]]

File: rust/glacier/src/widget/prompt.rs
//! Prompt widget.
//!
//! [`Prompt`] are [`TextInput`] area that call some callback upon submission. The primary usecase
//! is to spawn some program when the `Prompt` is submitted, but the default callback can be
//! overridden.
//!
//! By default, the prompt widget is hidden, and only shown when the [`Prompt`] is active by mean
//! of calling the [`activate`] function. Calling that function will display the prompt and request
//! focus to the underlying layer, so the [`Prompt`] gets all keyboard inputs. When `Enter` is
//! pressed, the prompt content will be submit and sent to the prompt [`exe_callback`], or
//! [`spawn`] if no callback was set.
//!
//! [`activate`]: Prompt::activate
//! [`exe_callback`]: Prompt::exe_callback

use std::marker::PhantomData;

use snowcap_api::{
    popup::Parent,
    widget::{
        Alignment, Border, Length, Padding,
        container::Container,
        font,
        text_input::{Icon, TextInput},
    },
};

use crate::{
    color,
    signal::WithEmitter,
    widget::{
        State, WeakState, Widget, WidgetMessage, WithState,
        base::WidgetBase,
        message::{self, MessageBuilder},
        operation, signal,
    },
};

pub mod style;
#[doc(inline)]
pub use style::{PromptStyle, Style};

/// [`Prompt`] messages type.
pub type Message = message::Message<Action>;
type ExeCallback = Box<dyn FnMut(&str) + Send + Sync>;

/// [`Prompt`] inner state.
pub struct Inner<Msg> {
    base: WidgetBase,
    placeholder: String,
    content: String,
    style: Style,
    exe_callback: Option<ExeCallback>,
    active: bool,
    id: String,
    message_builder: MessageBuilder<Action>,
    _msg: PhantomData<Msg>,
}

/// [`Prompt`] actions.
#[derive(Clone, Debug)]
pub enum Action {
    Input(String),
    Submit,
}

/// Prompt widget.
///
/// See module level [documentation] for more information.
///
/// [documentation]: self
#[derive(Clone)]
pub struct Prompt<Msg> {
    state: State<Inner<Msg>>,
}

/// Non-owning [`Prompt`].
#[derive(Clone)]
pub struct WeakPrompt<Msg>(WeakState<Inner<Msg>>);

/// Default [`Prompt`] execution callback.
pub fn spawn(input: &str) {
    if input.is_empty() {
        return;
    }

    let mut split = input.split_whitespace();

    let Some(cmd) = split.next() else {
        return;
    };

    pinnacle_api::process::Command::new(cmd).args(split).spawn();
}

/// Default [`Prompt`] appearance.
pub fn default_style() -> Style {
    Style::new()
        .font(
            font::Font::new()
                .family(font::Family::Monospace)
                .weight(font::Weight::Semibold),
        )
        .icon(Icon::new().code_point('').spacing(4.0))
        .bg_color(color::from_hex_alpha("#000000", 0.0))
        .border(Border {
            width: Some(0.),
            ..Default::default()
        })
        .padding(Padding::from(0.))
}

impl<Msg> Prompt<Msg>
where
    Msg: Clone + Send + Sync + 'static,
{
    const WIDGET_TYPE: &'static str = "Prompt";

    /// Create a new [`Prompt`]
    pub fn new() -> Self {
        let base = WidgetBase::new(Self::WIDGET_TYPE);
        let id = base.to_string();
        let message_builder = MessageBuilder::new(base.id());

        let style = default_style();

        let state = State::new(Inner {
            base,
            placeholder: String::default(),
            content: String::default(),
            style,
            exe_callback: None,
            active: false,
            id,
            message_builder,
            _msg: PhantomData,
        });

        Self { state }
    }

    /// Sets the [`Prompt`] style.
    pub fn style(self, style: Style) -> Self {
        self.state.0.lock().unwrap().style = style;
        self.emit(signal::RedrawNeeded);
        self
    }

    /// Deactivate the [`Prompt`].
    pub fn deactivate(&mut self) {
        self.state.0.lock().unwrap().deactivate();
    }

    /// Activate the [`Prompt`].
    pub fn activate(&mut self) {
        self.state.0.lock().unwrap().activate();
    }

    /// Remove the [`Prompt`] focus.
    pub fn unfocus(&mut self) {
        self.state.0.lock().unwrap().unfocus();
    }

    /// Sets the function to call when the prompt is submitted.
    pub fn exe_callback<F>(&mut self, callback: F)
    where
        F: FnMut(&str) + Send + Sync + 'static,
    {
        self.state.0.lock().unwrap().exe_callback = Some(Box::new(callback));
    }

    /// Create a [`WeakPrompt`] pointing to the same state.
    pub fn downgrade(&self) -> WeakPrompt<Msg> {
        WeakPrompt(self.state.downgrade())
    }
}

impl<Msg> WeakPrompt<Msg> {
    pub fn upgrade(&self) -> Option<Prompt<Msg>> {
        self.0.upgrade().map(|state| Prompt { state })
    }
}

impl<Msg> Inner<Msg> {
    fn refresh(&mut self) {
        self.emit(signal::RedrawNeeded);
    }

    fn focus(&mut self) {
        self.emit(signal::RequestFocus(self.id.clone()));
    }

    fn reset(&mut self) {
        self.content.clear();
    }

    fn unfocus(&mut self) {
        self.emit(signal::RequestUnfocus);
    }

    fn activate(&mut self) {
        if self.active {
            return;
        }

        self.active = true;
        self.refresh();
        self.focus();
    }

    fn deactivate(&mut self) {
        self.reset();
        self.active = false;

        self.refresh();
    }

    fn on_input(builder: MessageBuilder<Action>, input: String) -> WidgetMessage {
        builder.input(input)
    }
}

impl MessageBuilder<Action> {
    fn input(&self, input: String) -> WidgetMessage {
        self.build(Action::Input(input))
    }

    fn submit(&self) -> WidgetMessage {
        self.build(Action::Submit)
    }
}

impl<Msg> WithEmitter for Inner<Msg> {
    fn with_emitter(&self) -> crate::signal::Emitter {
        self.base.with_emitter()
    }
}

impl<Msg> WithEmitter for Prompt<Msg> {
    fn with_emitter(&self) -> crate::signal::Emitter {
        self.state.0.lock().unwrap().with_emitter()
    }
}

impl<Msg> Widget for Inner<Msg>
where
    Msg: Clone + From<WidgetMessage> + Into<Option<WidgetMessage>> + Send + Sync + 'static,
{
    type Message = Msg;

    fn view(&self) -> Option<snowcap_api::widget::WidgetDef<Self::Message>> {
        if !self.active {
            return None;
        }

        let mut inner = TextInput::new(&self.placeholder, &self.content)
            .id(&self.id)
            .on_input({
                let builder = self.message_builder;
                move |input| Self::on_input(builder, input).into()
            })
            .on_submit(self.message_builder.submit().into())
            .style(self.style.clone().into());

        inner.padding = self.style.padding;
        inner.font = self.style.font.clone();
        inner.icon = self.style.icon.clone();

        let prompt = Container::new(inner)
            .height(Length::Fill)
            .width(Length::Fill)
            .vertical_alignment(Alignment::Center);

        Some(prompt.into())
    }

    fn update(&mut self, msg: Self::Message, _parent: Option<Parent>) {
        use operation::{Focusable::Unfocus, Operation::Focusable};

        let Some(msg) = msg.into() else {
            return;
        };

        let action = match msg {
            WidgetMessage::Operation(Focusable(Unfocus)) => {
                self.deactivate();
                return;
            }
            WidgetMessage::Prompt(Message { id, action }) if id == self.base.id() => action,
            _ => {
                return;
            }
        };

        match action {
            Action::Input(s) => self.content = s,
            Action::Submit => {
                if let Some(callback) = &mut self.exe_callback {
                    callback(&self.content)
                } else {
                    spawn(&self.content);
                }

                self.unfocus()
            }
        }
    }
}

impl<Msg> Default for Prompt<Msg>
where
    Msg: Clone + Send + Sync + 'static,
{
    fn default() -> Self {
        Self::new()
    }
}

impl<Msg> WithState for Prompt<Msg> {
    type Type = Inner<Msg>;

    fn with_state(&self) -> State<Self::Type> {
        self.state.clone()
    }
}

impl From<Message> for WidgetMessage {
    fn from(value: Message) -> Self {
        Self::Prompt(value)
    }
}
